### PR TITLE
Enable Curve448 support via the PSA API

### PR DESCRIPTION
**Pull request by silabs-ArchanaM**
## Description
Enabled Curve448 support via the PSA API fixing #4249.
Added missing support for validating Curve448 keys fixing #3399. 

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Todos
- [x] Tests
- [x] Changelog updated

### CODE REVIEW

**Review by TRodziewicz — CHANGES_REQUESTED** [review]
I think having a changelog file would be good here.

File: tests/suites/test_suite_psa_crypto.data [hunk 1]
```diff
@@ -209,6 +209,22 @@ PSA import/export-public EC curve25519: accept masked input
 depends_on:PSA_WANT_ALG_ECDH:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:MBEDTLS_PK_PARSE_C:MBEDTLS_PK_WRITE_C:PSA_WANT_ECC_MONTGOMERY_255
 import_export_public_key:"70076d0a7318a57d3c16c17251b26645df4c2f87ebc0992ab177fba51db92c6a":PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_MONTGOMERY):PSA_ALG_ECDH:0:PSA_SUCCESS:"8520f0098930a754748b7ddcb43ef75a0dbf3a0d26381af4eba4a98eaa9b4e6a"
 
+PSA import/export EC curve448 key pair: good (already properly masked)
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_psa_crypto.data:212]:
```suggestion
PSA import/export EC curve448 key pair: good (already properly masked, key from RFC 7748 6.2 Alice)
```
To know where the private key comes from.

File: tests/suites/test_suite_psa_crypto.data [hunk 2]
```diff
@@ -209,6 +209,22 @@ PSA import/export-public EC curve25519: accept masked input
...
+depends_on:PSA_WANT_ALG_ECDH:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:MBEDTLS_PK_PARSE_C:MBEDTLS_PK_WRITE_C:PSA_WANT_ECC_MONTGOMERY_448
+import_export:"988f4925d1519f5775cf46b04b5800d4ee9ee8bae8bc5565d498c28dd9c9baf574a9419744897391006382a6f127ab1d9ac2d8c0a59872eb":PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_MONTGOMERY):PSA_KEY_USAGE_EXPORT:PSA_ALG_ECDH:448:0:PSA_SUCCESS:1
+
+PSA import/export EC curve448 key pair: unmasked input (check export-import-export yields properly masked output)
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_psa_crypto.data:216]:
```suggestion
PSA import/export EC curve448 key pair: unmasked input (check export-import-export yields properly masked output, key from RFC 7748 6.2 Alice)
```

File: tests/suites/test_suite_psa_crypto.data [hunk 3]
```diff
@@ -209,6 +209,22 @@ PSA import/export-public EC curve25519: accept masked input
...
+depends_on:PSA_WANT_ALG_ECDH:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:MBEDTLS_PK_PARSE_C:MBEDTLS_PK_WRITE_C:PSA_WANT_ECC_MONTGOMERY_448
+import_export:"9a8f4925d1519f5775cf46b04b5800d4ee9ee8bae8bc5565d498c28dd9c9baf574a9419744897391006382a6f127ab1d9ac2d8c0a598726b":PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_MONTGOMERY):PSA_KEY_USAGE_EXPORT:PSA_ALG_ECDH:448:0:PSA_SUCCESS:0
+
+PSA import/export-public EC curve448: accept masked input
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_psa_crypto.data:220]:
```suggestion
PSA import/export-public EC curve448: accept masked input (key from RFC 7748 6.2 Alice)
```

File: tests/suites/test_suite_psa_crypto.data [hunk 4]
```diff
@@ -209,6 +209,22 @@ PSA import/export-public EC curve25519: accept masked input
...
+depends_on:PSA_WANT_ALG_ECDH:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:MBEDTLS_PK_PARSE_C:MBEDTLS_PK_WRITE_C:PSA_WANT_ECC_MONTGOMERY_448
+import_export_public_key:"988f4925d1519f5775cf46b04b5800d4ee9ee8bae8bc5565d498c28dd9c9baf574a9419744897391006382a6f127ab1d9ac2d8c0a59872eb":PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_MONTGOMERY):PSA_ALG_ECDH:0:PSA_SUCCESS:"9b08f7cc31b7e3e67d22d5aea121074a273bd2b83de09c63faa73d2c22c5d9bbc836647241d953d40c5b12da88120d53177f80e532c41fa0"
+
+PSA import/export-public EC curve448: accept unmasked input
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_psa_crypto.data:224]:
```suggestion
PSA import/export-public EC curve448: accept unmasked input (key from RFC 7748 6.2 Alice)
```

File: tests/suites/test_suite_psa_crypto.data [hunk 5]
```diff
@@ -229,6 +245,10 @@ PSA import/export curve25519 public key: good
 depends_on:PSA_WANT_ALG_ECDH:PSA_WANT_KEY_TYPE_ECC_PUBLIC_KEY:MBEDTLS_PK_PARSE_C:MBEDTLS_PK_WRITE_C:PSA_WANT_ECC_MONTGOMERY_255
 import_export:"8520f0098930a754748b7ddcb43ef75a0dbf3a0d26381af4eba4a98eaa9b4e6a":PSA_KEY_TYPE_ECC_PUBLIC_KEY(PSA_ECC_FAMILY_MONTGOMERY):PSA_KEY_USAGE_EXPORT:PSA_ALG_ECDH:255:0:PSA_SUCCESS:1
 
+PSA import/export curve448 Public Key: good
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_psa_crypto.data:248]:
```suggestion
PSA import/export curve448 Public Key: good (key from RFC 7748 6.2 Alice)
```

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
That looks almost good to me, thanks. I have only minor comments. I would also suggest to add a "PSA generate key: ECC, Curve448, good" test similar to the "PSA generate key: ECC, Curve25519, good" one to complete the alignment of curve 25519 and curve 448 testing in test_sute_psa_crypto. Otherwise, the PR needs rebasing now.

File: library/ecp.c [hunk 6]
```diff
@@ -3276,12 +3292,14 @@ int mbedtls_ecp_write_key( mbedtls_ecp_keypair *key,
...
 
-            MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary_le( &key->d, buf, buflen ) );
         }
-        else
-            ret = MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE;
+        else if ( key->grp.id == MBEDTLS_ECP_DP_CURVE448 )
+        {
+            if( buflen < ECP_CURVE448_KEY_SIZE )
+                return MBEDTLS_ERR_ECP_BUFFER_TOO_SMALL;
+        }
```
[the remark this answers is not in the document]
Comment (silabs-ArchanaM) [reply library/ecp.c:3300]: The default return value is already ret = MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
Thanks for addressing my comments so quickly. There are a few things that we use to do differently though:
1) when a PR is in review and it has some conflicts with the branch it is targeted to we prefer to rebase the PR rather than merge the targeted branch (development in the case of this PR). Thus, could you please remove the last merge commit and rather rebase the PR commits on top of the `development` branch. Feel free to ask if you have some doubt on how to do that.
2) concerning review conversation we let the reviewer not the PR author resolve the conversations. The PR author usually just uses the rocket icon to indicate to the reviewer that a comment has been addressed.
3) when you create new commits to address review comments please make commit message that explain what the commit does and group in one commit similar changes. For example, here a possibility is to have one commit to address the style issues with a commit message like "Fix coding style issue" (it would a squash of your three "Update library/ecp.c" commits. Then we could have one commit "test: psa: Improve the description of some curve 448 tests" ...

File: library/ecp.c [hunk 7]
```diff
@@ -3235,8 +3236,23 @@ int mbedtls_ecp_read_key( mbedtls_ecp_group_id grp_id, mbedtls_ecp_keypair *key,
...
+            /* Set the two least significant bits to 0 */
+            MBEDTLS_MPI_CHK( mbedtls_mpi_set_bit( &key->d, 0, 0 ) );
+            MBEDTLS_MPI_CHK( mbedtls_mpi_set_bit( &key->d, 1, 0 ) );
+
+            /* Set the most significant bit to 1 */
+            MBEDTLS_MPI_CHK(
+                    mbedtls_mpi_set_bit( &key->d,
+                                         ECP_CURVE448_KEY_SIZE * 8 - 1, 1 )
+                    );
+        }
```
[the remark this answers is not in the document]
Comment (silabs-ArchanaM) [reply library/ecp.c:3120]: The default return value is already ret = MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
Thanks for the changes. Almost good to me, please my last comments below.

**Review by ronald-cron-arm — COMMENTED** [review]
Thanks for the changes. You have some trailing spaces in the change log and the CI is not happy. Otherwise LGTM.

**Review by ronald-cron-arm — APPROVED** [review]
LGTM, thanks.

**Review by TRodziewicz — APPROVED** [review]

File: ChangeLog.d/psa_curve448_key_support.txt [hunk 8]
```diff
@@ -0,0 +1,3 @@
+Features
+   * Enable support for Curve448 via the PSA API, thereby fixing #4249 and
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply ChangeLog.d/psa_curve448_key_support.txt]: A while back we changed our policy to not acknowledge contributors in the changelog anymore. The acknowledgement policy dated back from when the change history was private. Now contributors are acknowledged via the public Git history.

Not a blocker, but please don't feel slighted if we remove this when we edit the changelog for the next release.

**Pull request merged**